### PR TITLE
fix(api): remove unnecessary pipette movement during mix

### LINKED ISSUE #4640: APIv2: pipette moves between mix() repetitions
# Current behavior

Between aspirate-dispense repetitions within a single `mix()` command, the pipette tip will move to the top of the well and then immediately move back down.

# Expected behavior

The pipette tip should stay at the same height for the entire `mix()` command.

# APIv2 only

This behavior is *not* present when uploading an APIv1 protocol. In APIv1, the pipette stays at the same height, as expected.

# Test case

```
metadata = {
    'apiLevel': '2.0'
}

def run(protocol):
    plate = protocol.load_labware('corning_96_wellplate_360ul_flat', '2')
    tiprack = protocol.load_labware('opentrons_96_tiprack_300ul', '1')
    pipette = protocol.load_instrument('p300_multi', 'right', tip_racks=[tiprack])
    pipette.pick_up_tip()
    pipette.mix(repetitions=4, volume=150, location=plate.wells_by_name()['A1'])
    pipette.return_tip()
```

# See also

- #780 maybe?

### PR DESCRIPTION
**Pull request by ahiuchingau**
## overview
In APIv2, the pipette would move above the well between each aspirate-dispense repetitions during mix. The original logic for this movement is to reset the plunger position above liquid level after a blow-out command, to avoid over-aspiration in the subsequent `aspirate`. 

However, we currently evaluate the necessity of this action by checking the current volume of the pipette: if the volume is equal to 0, the pipette is always moved above the top of the well. Instead, we should check the plunger position and only move the pipette when the position is below the `bottom` position.

closes #4640 

## changelog
- add a check for the plunger position in `aspirate` to avoid unnecessary pipette movement
- include the pipette bottom position in the pipette config dictionary for easy access in the protocol API
- add function in hardware control API to return the current plunger position
- bump max API version to 2.3

## review requests
Run the following protocol in API version 2.2 and 2.3 and see that:
- in v2.2, the pipette moves out of the well before each aspirate
- in v2.3, the pipette only moves out of the well after blowout
```
pip.pick_up_tip()
pip.mix(repetitions=3)
pip.blow_out()
pip.mix(repetitions=3)
pip.return_tip()
```

### CODE REVIEW

File: api/src/opentrons/protocol_api/instrument_context.py [hunk 1]
```diff
@@ -106,6 +106,11 @@ def reset_tipracks(self):
             tiprack.reset()
         self.starting_tip = None
 
+    @property  # type: ignore
+    @requires_version(2, 3)
```
Comment (sfoster1) [inline api/src/opentrons/protocol_api/instrument_context.py]: underscore methods, which are private by convention, don't need requires_version annotations

File: api/src/opentrons/hardware_control/api.py [hunk 2]
```diff
@@ -604,6 +604,21 @@ def _deck_from_smoothie(
                 plunger_ax: self._current_position[plunger_ax]
             }
 
+    async def plunger_position(
```
Comment (sfoster1) [inline api/src/opentrons/hardware_control/api.py]: This doesn't feel like quite the right interface - the hardware controller's abstraction of a pipette is based around volumes, not positions. 

`hardware_control.pipette.Pipette` actually already has a `ready_to_aspirate` flag that is managed by `hardware_control.API.prepare_for_aspirate`. If we expose that through the `hardware_pipette` interface, we can gate both the move to top and, if we want, the call to `prepare_for_aspirate` behind it.

File: api/tests/opentrons/protocol_api/test_context.py [hunk 3]
```diff
@@ -370,14 +367,11 @@ def test_aspirate(loop, get_labware_def, monkeypatch):
...
     dest_point, dest_lw = lw.wells()[0].bottom()
     dest_point = dest_point._replace(z=dest_point.z + 1.0)
-    assert fake_move.call_args_list[-2] ==\
-        mock.call(Mount.RIGHT, lw.wells()[0].top().point,
-                  critical_point=None, speed=400, max_speeds={})
     assert fake_move.call_args_list[-1] ==\
         mock.call(
             Mount.RIGHT, dest_point, critical_point=None, speed=400,
             max_speeds={})
-    assert len(fake_move.call_args_list) == 2
```
Comment (sfoster1) [inline api/tests/opentrons/protocol_api/test_context.py]: we should add a test for the case where we do in fact want to prepare for aspirate here

**Review by sfoster1 — CHANGES_REQUESTED** [review]
Definitely the right way to solve it in general, but some of this functionality is actually in `hardware_control.pipette.Pipette` already, let's use that

File: api/src/opentrons/hardware_control/api.py (hunk 2 above)
Comment (sfoster1) [reply api/src/opentrons/hardware_control/api.py]: delete

**Review by sfoster1 — APPROVED** [review]
Code LGTM once we delete the now-unused `plunger_position` function, feel free to merge once @nusrat813 approves

**Comment by nusrat813** [comment]
Issues with aspirate and dispense - not using max volume if volume not specified

**Review by sfoster1 — APPROVED** [review]
LGTM after publish changes

**Comment by nusrat813** [comment]
Retested after publish changes, working as intended 👍

**Review by nusrat813 — APPROVED** [review]

**Comment by SyntaxColoring** [comment]
I believe we need documentation and API changelog updates as well, right?

**Pull request merged**